### PR TITLE
Prefer python meta function over c++ meta function

### PR DESCRIPTION
**Pull request by SherlockNoMad**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #87662
* #87563
* #87569
* __->__ #87426

This is a policy update for meta registration. **We now prefer python meta implementation over C++ meta function.**  This is a flip of the previous policy, where we prefer C++ meta function over python meta function if they both exist. 

Here's the meta registration process:
1. register_meta and register_decomposition will place the python meta/decomp functions into the `global_decomp_table`.  However, they will NOT register them into dispatcher. 
2. After global_decomp_table is populated, we will compile an `active_meta_table`. For a given op, we pick the most specific decomp function from `global_decomp_table` in the preference order of Meta > PostAutograd > PreAutograd. 
3. We will unconditionally register all of them into python dispatcher. And register them into C++ dispatcher, unless it one of the following 3 cases
- 1. the op is a CompositeImplicitAutograd, and should rely on decomposed op's meta
- 2. the op is a view op, as the MetaTensor doesn't support aliased storage
- 3. the op is in the blocklist (due to UT failures, and we will burn down this list op by op) 

Over the long run, we wish to implement all meta functions in python. With this PR, 321 op_overloads will have cpp meta overridden by python meta. There are still 400 op_overloads is using cpp meta. The exact list can be found here https://gist.github.com/SherlockNoMad/d20bb736178df8eebd3b054c8bb7cdc5

cc @ngimel @jansel @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @lezcano @fdrocha

### CODE REVIEW

File: torch/_meta_registrations.py [hunk 1]
```diff
@@ -1636,3 +1638,64 @@ def upsample_nearest2d_vec(input, output_size, scale_factors):
...
+            "aten::empty.names",  # TypeError: empty() got an unexpected keyword argument 'names', test/inductor/test_torchinductor.py -k test_zeros_cpu
+            "aten::add.Tensor",  # ValueError: Receive two Number inputs to an elementwise binary operation! test/inductor/test_torchinductor.py -k test_both_scalars
+            "aten::sub.Tensor",  # ValueError: Receive two Number inputs to an elementwise binary operation! test/inductor/test_torchinductor.py -k test_both_scalars
+            "aten::mul.Tensor",  # ValueError: Receive two Number inputs to an elementwise binary operation! test/inductor/test_torchinductor.py -k test_both_scalars
+            "aten::div.Tensor",  # ValueError: Receive two Number inputs to an elementwise binary operation! test/test_fake_tensor.py -k test_scalar_inputs
+            "aten::div.Tensor_mode",  # ValueError: Receive two Number inputs to an elementwise binary operation! test/inductor/test_torchinductor.py -k test_div8_cpu
+            "aten::diag_embed",  # RuntimeError: Stride mismatch! Strides are (180, 30, 1, 6) and (180, 30, 5, 1) (mismatched at 2)! test/test_ops.py -k test_fake_autocast_diag_embed_cuda_float32
```
Comment (eellison) [inline torch/_meta_registrations.py:1692]: it seems like we're just moving the public api into a private list. Is that useful ?

**Review by eellison — COMMENTED** [review]
Couple comments/questions

File: torch/_decomp/decompositions.py [hunk 2]
```diff
@@ -1929,7 +1929,7 @@ def is_same_size(a: Tensor, b: Tensor) -> bool:
     return a.shape == b.shape
 
 
-@register_decomposition([aten._reshape_alias, aten._unsafe_view], disable_meta=True)
+@register_decomposition([aten._reshape_alias, aten._unsafe_view])
 def _reshape_alias(x, shape, *args):
     return aten.view(x, shape)
```
[the remark this answers is not in the document]
Comment (SherlockNoMad) [reply torch/_decomp/decompositions.py:1934]: _reshape_alias is not registered to C++ dispatcher, as it's a view op. See the registration rule in the description.

But yes, we should still fix the stride. I am wondering if we can track this by asserting the stride correctness in test_meta.py.

[the remark this answers is not in the document]
Comment (Chillee) [reply torch/_decomp/decompositions.py:1934]: I don't see why that's incorrect? From my understanding, `_reshape_alias` is just `view` but with the strides precomputed.

https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/TensorShape.cpp#L1419

image

File: torch/_meta_registrations.py (hunk 1 above)
Comment (SherlockNoMad) [reply torch/_meta_registrations.py:1692]: This is a list I will work to burn down. 
Most of them can be resolved by fixing the meta implementation. 

For the ones that python meta doesn't work, I plan to add it to the block list in functions_with_cpp_meta_impl_that_support_symint. 

So eventually, it should just look up from functions_with_cpp_meta_impl_that_support_symint.

File: torch/_meta_registrations.py [hunk 3]
```diff
@@ -1636,3 +1638,64 @@ def upsample_nearest2d_vec(input, output_size, scale_factors):
...
+        registry = global_decomposition_table[type]
+
+        for opo in registry:
+            if opo not in activate_meta_table:
+                activate_meta_table[opo] = registry[opo]
+
+    for op_overload, fn in activate_meta_table.items():
+        assert isinstance(op_overload, OpOverload)
+
+        op_overload.py_impl(torch._C.DispatchKey.Meta)(fn)
```
Comment (SherlockNoMad) [inline torch/_meta_registrations.py:1657]: @ezyang, 
I am wondering if we should make python dispatcher following the registration logic as c++ dispatcher. 

If yes, we will have the consistent behavior between python dispatcher and c++ dispatcher on Meta key.

File: torch/_decomp/decompositions.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (eellison) [reply torch/_decomp/decompositions.py:1934]: https://codebrowser.bddppq.com/pytorch/pytorch/aten/src/ATen/native/TensorShape.cpp.html#_ZN2at6native14_reshape_aliasERKNS_6TensorEN3c108ArrayRefIlEES6_


```
>>> x = torch.rand([8, 4])
>>> y = torch.ops.aten._reshape_alias(x, [8, 4], [1, 4])
>>> y.stride()
(1, 4)
>>> x.view([8, 4]).stride()
(4, 1)
>>> 
```

If the strides are always pre-computed to be consistent we should assert that.

File: aten/src/ATen/core/dispatch/OperatorEntry.cpp [hunk 4]
```diff
@@ -147,13 +147,15 @@ OperatorEntry::AnnotatedKernelContainerIterator OperatorEntry::registerKernel(
...
   if (k.size() > 0) {
 #endif
-    TORCH_WARN("Overriding a previously registered kernel for the same operator and the same dispatch key\n",
-               "  operator: ", (schema_.has_value() ? toString(schema_->schema) : toString(name_)), "\n",
-               "    ", (this->schema_.has_value() ? this->schema_->debug : "no debug info"), "\n",
-               "  dispatch key: ", toString(dispatch_key), "\n",
-               "  previous kernel: ", (cpp_signature_.has_value() ? cpp_signature_->debug : (sym_cpp_signature_.has_value() ? sym_cpp_signature_->debug : "no debug info")), "\n",
-               "       new kernel: ", debug
-    );
+    if (dispatch_key != DispatchKey::Meta) {
```
Comment (ezyang) [inline aten/src/ATen/core/dispatch/OperatorEntry.cpp:150]: Good to have a comment why Meta is suppressed specifically

Comment (SherlockNoMad) [reply aten/src/ATen/core/dispatch/OperatorEntry.cpp:152]: done.

File: test/test_dispatch.py [hunk 5]
```diff
@@ -765,6 +765,7 @@ def test_overwrite_math(self):
     # no longer exists
     def test_find_dangling_impls(self):
         dangling_impls = C._dispatch_find_dangling_impls()
+        print(dangling_impls)
```
Comment (SherlockNoMad) [inline test/test_dispatch.py:768]: removed.

File: test/test_functionalization.py [hunk 6]
```diff
@@ -1033,8 +1033,8 @@ def forward(self, a_1):
     add = torch.ops.aten.add.Tensor(a_1, 1);  a_1 = None
     resize = torch.ops.aten.resize_.default(add, [5, 5])
     view = torch.ops.aten.view.default(add, [25]);  add = None
-    fill = torch.ops.aten.fill_.Scalar(view, 1)
-    view_1 = torch.ops.aten.view.default(view, [5, 5]);  view = None
+    fill = torch.ops.aten.fill.Scalar(view, 1);  view = None
+    view_1 = torch.ops.aten.view.default(fill, [5, 5]);  fill = None
```
Comment (ezyang) [inline test/test_functionalization.py:1037]: This is a regression , cc @bdhirsh

File: test/test_ops.py [hunk 7]
```diff
@@ -1909,6 +1908,8 @@ def test_refs_are_in_decomp_table(self, op):
     skip("linalg.pinv", "hermitian"),
     skip("linalg.pinv", "singular"),
     skip('pinverse'),
+    skip('cholesky_inverse'),  # test/test_ops.py -k test_fake_crossref_backward_amp_cholesky_inverse_cuda_float32
```
Comment (ezyang) [inline test/test_ops.py:1911]: this failing now is weird

File: torch/_subclasses/fake_tensor.py [hunk 8]
```diff
@@ -780,7 +780,6 @@ def __torch_dispatch__(self, func, types, args=(), kwargs=None):
         if (
             func in decomposition_table
             and torch_decomp_decompositions(func)
-            and func not in _disabled_meta_decomps
```
Comment (ezyang) [inline torch/_subclasses/fake_tensor.py:783]: @eellison check this

File: torch/_meta_registrations.py [hunk 9]
```diff
@@ -1636,3 +1629,60 @@ def upsample_nearest2d_vec(input, output_size, scale_factors):
...
 import torch._refs.nn.functional
 import torch._refs.special
+
+
+def activate_meta():
+
+    activate_meta_table = {}
+
+    for type in ["meta", "post_autograd", "pre_autograd"]:
+        registry = global_decomposition_table[type]
```
Comment (ezyang) [inline torch/_meta_registrations.py:1639]: why not just iterate over the keys of the table

Comment (ezyang) [reply torch/_meta_registrations.py:1639]: oh nvm

File: torch/_meta_registrations.py [hunk 10]
```diff
@@ -1636,3 +1629,60 @@ def upsample_nearest2d_vec(input, output_size, scale_factors):
 import torch._refs
 import torch._refs.nn.functional
 import torch._refs.special
+
+
+def activate_meta():
+
+    activate_meta_table = {}
+
+    for type in ["meta", "post_autograd", "pre_autograd"]:
```
Comment (ezyang) [inline torch/_meta_registrations.py:1638]: comment something about precedence order

File: torch/_meta_registrations.py [hunk 11]
```diff
@@ -1636,3 +1629,60 @@ def upsample_nearest2d_vec(input, output_size, scale_factors):
...
+        for opo in registry:
+            if opo not in activate_meta_table:
+                activate_meta_table[opo] = registry[opo]
+
+    for op_overload, fn in activate_meta_table.items():
+        assert isinstance(op_overload, OpOverload)
+
+        op_overload.py_impl(torch._C.DispatchKey.Meta)(fn)
+
+        dispatch_key_registration = torch._C._dispatch_dump(op_overload.name())
```
Comment (ezyang) [inline torch/_meta_registrations.py:1650]: don't do it this way; use one of the more specific dispatch query functions

**Review by ezyang — APPROVED** [review]
some small nits, looks good, much simpler

File: test/test_functionalization.py (hunk 6 above)
Comment (bdhirsh) [reply test/test_functionalization.py:1037]: probably a bug in one of our (related) python decomps?

File: torch/_meta_registrations.py (hunk 11 above)
Comment (SherlockNoMad) [reply torch/_meta_registrations.py:1650]: Shall I also check for CompositeImplicitAutogradNestedTensor? 

aten.unbind has this Dispatchkey.

Comment (SherlockNoMad) [reply torch/_meta_registrations.py:1650]: I suppose it's not needed? 

For non-nested tensor, aten.unbind should still depends on python's meta.

Comment (SherlockNoMad) [reply torch/_meta_registrations.py:1650]: uhhh... library.py is checking CompositeImplicitAutogradNestedTensor


I am fix this in the next PR.

File: test/test_functionalization.py (hunk 6 above)
Comment (SherlockNoMad) [reply test/test_functionalization.py:1037]: fixed. 
It's a prim's implementation bug.

File: torch/_meta_registrations.py (hunk 10 above)
Comment (SherlockNoMad) [reply torch/_meta_registrations.py:1638]: done.

File: test/test_functionalization.py (hunk 6 above)
Comment (SherlockNoMad) [reply test/test_functionalization.py]: Fixed here https://github.com/pytorch/pytorch/pull/87493

File: torch/_meta_registrations.py (hunk 3 above)
Comment (SherlockNoMad) [reply torch/_meta_registrations.py:1647]: OK, will do in next PR.

File: test/test_ops.py [hunk 12]
```diff
@@ -1909,6 +1908,7 @@ def test_refs_are_in_decomp_table(self, op):
...
```
Comment (eellison) [inline test/test_ops.py:1911]: Would prefer if we didn't regress the tests to land this - could you look into this ?

File: torch/_subclasses/fake_tensor.py (hunk 8 above)
Comment (eellison) [reply torch/_subclasses/fake_tensor.py:783]: @SherlockNoMad, can you find to expose the set of ops you've disabled for meta decomps to not be called ? (previously `_disabled_meta_decomps`) ?

Comment (SherlockNoMad) [reply torch/_subclasses/fake_tensor.py:783]: Here's the exhausted list. Essentially, all the decomps previous marked with "disable_meta=True" will be in this list. 

```
aten::split_with_sizes
aten::split.Tensor
aten::native_group_norm
aten::detach
aten::lift
aten::lift.out
aten::lift_fresh
aten::_adaptive_avg_pool2d
aten::_adaptive_avg_pool2d.out
aten::_reshape_alias
aten::_unsafe_view
aten::_unsafe_view.out
aten::dot
aten::dot.out
aten::expand
aten::permute
aten::index_select
aten::index_select.out
aten::index_select.dimname
aten::index_select.dimname_out
aten::squeeze
aten::squeeze.dim
aten::squeeze.dimname
aten::diagonal
aten::diagonal.Dimname
aten::t
aten::transpose.int
aten::transpose.Dimname
aten::unsqueeze
aten::view
aten::view.dtype
```

Comment (eellison) [reply torch/_subclasses/fake_tensor.py:783]: Yea, we can we avoid calling those decopmositions here in fake tensor, as we were before ?

File: torch/_meta_registrations.py (hunk 11 above)
Comment (ezyang) [reply torch/_meta_registrations.py:1649]: for the CompositeImplicitAutograd test you just check if the (non-computed) key is registered

File: test/test_ops.py (hunk 12 above)
Comment (SherlockNoMad) [reply test/test_ops.py]: Fixed. It was flaky, and not failing now...

File: test/test_ops.py (hunk 7 above)
Comment (SherlockNoMad) [reply test/test_ops.py]: Fixed. It was flaky, and not failing now...

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**